geckocamera.cpp

```diff
@@ -42,37 +43,27 @@ class RootCameraManager : public CameraManager
     bool openCamera(const string cameraId, shared_ptr<Camera> &camera);
 
 private:
-    bool initialized = false;
     void findCameras();
+    shared_ptr<CameraManager> loadPlugin(Plugin &plugin);
 
-    vector<CameraInfo> cameraInfoList;
-    map<const string, shared_ptr<CameraManager>> cameraIdMap;
-    map<const string, shared_ptr<CameraManager>> plugins;
-
-    shared_ptr<CameraManager> loadPlugin(string path);
+    mutex m_mutex;
+    bool m_initialized = false;
+    vector<CameraInfo> m_cameraInfoList;
+    map<const string, shared_ptr<CameraManager>> m_cameraIdMap;
+    map<const string, shared_ptr<CameraManager>> m_plugins;
 };
 
 bool RootCameraManager::init()
 {
-    if (!initialized) {
-        LogInit("gecko-camera", getenv("GECKO_CAMERA_DEBUG") ? LogDebug : LogInfo);
-        filesystem::directory_entry pluginDir(GECKO_CAMERA_PLUGIN_DIR);
-        if (pluginDir.exists() && pluginDir.is_directory()) {
-            for (const auto &entry : filesystem::directory_iterator(GECKO_CAMERA_PLUGIN_DIR)) {
-                if (entry.is_regular_file()) {
-                    string path = entry.path();
-                    auto found = plugins.find(path);
-                    if (found == plugins.end()) {
-                        auto plugin = loadPlugin(path);
-                        if (plugin && plugin->init()) {
-                            LOGI("Initialized plugin at " << path);
-                            plugins.emplace(path, plugin);
-                        }
-                    }
-                }
+    scoped_lock lock(m_mutex);
+    if (!m_initialized) {
+        for (auto plugin : PluginManager::get()->listPlugins()) {
+            auto manager = loadPlugin(plugin);
+            if (manager && manager->init()) {
+                m_plugins.emplace(plugin.path, manager);
             }
         }
-        initialized = true;
+        m_initialized = true;
     }
     return true;
```

Comment:
Is it correct that this returns `true` even if it's already been initialised? I only ask because the behaviour differs from that of `DroidVideoEncoder::init()`, which returns `false` if called multiple times.

Is it also correct that initialisation completes successfully, even if `manager->init()` returns false?

Reply:
I didn't see any change here, but if it's correct to be like this, then it's fine for me too.

Reply:
I not pushed my change by mistake but I'm starting to remember that I kind of made this intentionally. If RootCameraManager didn't find any plugins it will retry next time.

Reply:
I not pushed my change by mistake, but after a thought I think this is correct code, `RootCameraManager::init()` always returns `true`, it cannot fail.

Reply:
Perfect, thank you.